// @jupyterlab/codeeditor/lib/widget.js
"use strict";
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var widgets_1 = require("@phosphor/widgets");
/**
 * The class name added to an editor widget that has a primary selection.
 */
var HAS_SELECTION_CLASS = 'jp-mod-has-primary-selection';
/**
 * A widget which hosts a code editor.
 */
var CodeEditorWrapper = (function (_super) {
    __extends(CodeEditorWrapper, _super);
    /**
     * Construct a new code editor widget.
     */
    function CodeEditorWrapper(options) {
        var _this = _super.call(this) || this;
        var editor = _this.editor = options.factory({
            host: _this.node,
            model: options.model,
            uuid: options.uuid,
            config: options.config,
            selectionStyle: options.selectionStyle
        });
        editor.model.selections.changed.connect(_this._onSelectionsChanged, _this);
        return _this;
    }
    Object.defineProperty(CodeEditorWrapper.prototype, "model", {
        /**
         * Get the model used by the widget.
         */
        get: function () {
            return this.editor.model;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Dispose of the resources held by the widget.
     */
    CodeEditorWrapper.prototype.dispose = function () {
        if (this.isDisposed) {
            return;
        }
        _super.prototype.dispose.call(this);
        this.editor.dispose();
    };
    /**
     * Handle `'activate-request'` messages.
     */
    CodeEditorWrapper.prototype.onActivateRequest = function (msg) {
        this.editor.focus();
    };
    /**
     * A message handler invoked on an `'after-attach'` message.
     */
    CodeEditorWrapper.prototype.onAfterAttach = function (msg) {
        _super.prototype.onAfterAttach.call(this, msg);
        if (this.isVisible) {
            this.editor.refresh();
        }
    };
    /**
     * A message handler invoked on an `'after-show'` message.
     */
    CodeEditorWrapper.prototype.onAfterShow = function (msg) {
        this.editor.refresh();
    };
    /**
     * A message handler invoked on a `'resize'` message.
     */
    CodeEditorWrapper.prototype.onResize = function (msg) {
        if (msg.width >= 0 && msg.height >= 0) {
            this.editor.setSize(msg);
        }
        else {
            this.editor.resizeToFit();
        }
    };
    /**
     * Handle a change in model selections.
     */
    CodeEditorWrapper.prototype._onSelectionsChanged = function () {
        var _a = this.editor.getSelection(), start = _a.start, end = _a.end;
        if (start.column !== end.column || start.line !== end.line) {
            this.addClass(HAS_SELECTION_CLASS);
        }
        else {
            this.removeClass(HAS_SELECTION_CLASS);
        }
    };
    return CodeEditorWrapper;
}(widgets_1.Widget));
exports.CodeEditorWrapper = CodeEditorWrapper;
